Aplib.Core/Belief/Beliefs/Belief.cs

```diff
@@ -30,57 +30,88 @@ public class Belief<TReference, TObservation> : IBelief where TReference : class
+            Func<bool> shouldUpdate
+        )
         {
             Type referenceType = reference.GetType();
             if (referenceType.IsValueType)
                 throw new ArgumentException($"{referenceType.FullName} is not a reference type.", nameof(reference));
 
+            Metadata = metadata;
             _reference = reference;
             _getObservationFromReference = getObservationFromReference;
             Observation = _getObservationFromReference(_reference);
+            _shouldUpdate = shouldUpdate;
         }
 
-        /// <summary>
-        /// Initializes a new instance of the <see cref="Belief{TReference, TObservation}"/> class with an object reference,
-        /// a function to generate/update the observation using the object reference,
-        /// and a condition on when the observation should be updated.
-        /// </summary>
-        /// <param name="reference">
-        /// The object reference used to generate/update the observation. This <i>must</i> be a reference type, be aware
-        /// that this is not enforced by C# if <typeparamref name="TReference"/> is an interface.
-        /// </param>
-        /// <param name="getObservationFromReference">A function that takes an object reference and generates/updates an observation.</param>
-        /// <param name="shouldUpdate">A condition on when the observation should be updated.</param>
-        /// <exception cref="ArgumentException">
-        /// Thrown when <paramref name="reference"/> is not a reference type.
-        /// </exception>
-        public Belief(TReference reference,
+        /// <inheritdoc
+        ///     cref="Belief{TReference,TObservation}(Aplib.Core.Metadata,TReference,Func{TReference,TObservation},Func{bool})"/>
+        public Belief
+        (
+            TReference reference,
             Func<TReference, TObservation> getObservationFromReference,
-            Func<bool> shouldUpdate)
-            : this(reference, getObservationFromReference)
+            Func<bool> shouldUpdate
+        )
+            : this(new Metadata(), reference, getObservationFromReference, shouldUpdate)
+        {
+        }
+
+        /// <inheritdoc
+        ///     cref="Belief{TReference,TObservation}(Aplib.Core.Metadata,TReference,Func{TReference,TObservation},Func{bool})" />
+        public Belief
+        (
+            Metadata metadata,
+            TReference reference,
+            Func<TReference, TObservation> getObservationFromReference
+        )
+            : this(metadata, reference, getObservationFromReference, () => true)
+        {
+        }
+
+        /// <inheritdoc
+        ///     cref="Belief{TReference,TObservation}(Aplib.Core.Metadata,TReference,Func{TReference,TObservation},Func{bool})" />
+        public Belief(TReference reference, Func<TReference, TObservation> getObservationFromReference)
+            : this(new Metadata(), reference, getObservationFromReference)
```

Comment:
I would try to prevent too much constructor chaining. It results in a little bit of code duplication, but if we don't do it, then
- The reader would need to move all around the file to find all the constructors that are being called. 
- There would be more function calls (not really a reason to prevent it per se, but a small extra con).
- The following scenario would not have a clear implementation:
```cs
public Foo(Bar a, Baz b, Qux c) { // Do stuff with a, b, c. }
public Foo(Bar a, Baz b) : this(a, b, null) { }
public Foo(Bar a, Qux c) : this(a, null, c) { }
public Foo(Bar a) : this(a, null) { } // Ambiguous. Also, if we would not give null, but an actual unambiguous value for the second parameter, then we would be able to choose whether we give a Baz or a Qux; there would be no reason to choose one over the other.
```
```suggestion
            : this(new Metadata(), reference, getObservationFromReference, () => true)
```